src/test/java/fi/aalto/cs/apluscourses/presentation/MainViewModelUpdaterTest.java

```diff
@@ -0,0 +1,139 @@
+import static org.mockito.Mockito.when;
+
+import com.intellij.openapi.module.Module;
+import com.intellij.openapi.module.ModuleManager;
+import com.intellij.openapi.project.Project;
+import com.intellij.testFramework.HeavyPlatformTestCase;
+import fi.aalto.cs.apluscourses.TestHelper;
+import fi.aalto.cs.apluscourses.intellij.model.APlusProject;
+import fi.aalto.cs.apluscourses.intellij.services.PluginSettings;
+import fi.aalto.cs.apluscourses.model.Course;
+import fi.aalto.cs.apluscourses.model.ModelExtensions;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.jetbrains.annotations.NotNull;
+import org.junit.Test;
+
+public class MainViewModelUpdaterTest extends HeavyPlatformTestCase implements TestHelper {
+
+  @Test
+  public void testGetProjectModuleNamesWithCorrectModulesReturnsValid() {
+    //  given
+    Project project = getProject();
+    createAndAddModule(project, "module1", "UID#1");
+    //  modules are identified by name, so only one is actually created
+    createAndAddModule(project, "module1", "UID#1");
+    createAndAddModule(project, "module2", "UID#1");
+    Module[] modules = ModuleManager.getInstance(project).getModules();
+    assertEquals("Only one module named 'module1' is created.", 3, modules.length);
+
+    MainViewModel mainViewModel = PluginSettings.getInstance().getMainViewModel(project);
+    MainViewModelUpdater mainViewModelUpdater = new MainViewModelUpdater(mainViewModel,
+        project, 1000L);
+
+    //  when
+    Set<String> projectModules = mainViewModelUpdater.getProjectModuleNames();
+
+    //  then
+    assertEquals("There are 3 (three) module names listed.", 3, projectModules.size());
+  }
+
+  @Test
+  public void testGetProjectModuleNamesWithClosedProjectReturnsEmpty() {
+    //  given
+    Project project = getProject();
+    Project spyProject = spy(project);
+    when(spyProject.isOpen()).thenReturn(false);
+
+    MainViewModel mainViewModel = PluginSettings.getInstance().getMainViewModel(spyProject);
+    MainViewModelUpdater mainViewModelUpdater = new MainViewModelUpdater(mainViewModel,
+        spyProject, 1000L);
+
+    //  when
+    Set<String> projectModules = mainViewModelUpdater.getProjectModuleNames();
+
+    //  then
+    assertEmpty("There are 0 (zero) module names listed for a closed project.", projectModules);
+  }
+
+  @Test
+  public void testAddSameModuleTwiceFails() throws MalformedURLException {
+    //  given, when, then, whatever
+    Course course = getDummyCourse();
+    /**
+     *  Check {@link MainViewModelUpdaterTest#getDummyCourse()} method.
+     **/
+    fi.aalto.cs.apluscourses.model.Module module2 = new ModelExtensions.TestModule("Module2");
+    assertThrows(UnsupportedOperationException.class, () -> course.getModules().add(module2));
+  }
```

Comment:
But for sure we can't have modules with duplicate names (and what weird course would have modules with the same name), because the local course file uses module names as keys, but I'm not sure if that's testable.

Reply:
this will fail even earlier @nikke234 ([it seems](https://github.com/Aalto-LeTech/intellij-plugin/blob/f32d04cb8043ee6874eece41006469714463bad8/src/test/java/fi/aalto/cs/apluscourses/model/CourseTest.java#L76))